commit 2b8ac062f33781edbefb482c6abebfde6b373eed
Author: Vincent Whitchurch <vincent.whitchurch@axis.com>
Date:   Thu Jul 1 10:05:34 2021 +0200

    mmc: dw_mmc: Add data CRC error injection
    
    This driver has had problems when handling data errors.  Add fault
    injection support so that the abort handling can be easily triggered and
    regression-tested.  A hrtimer is used to indicate a data CRC error at
    various points during the data transfer.
    
    Note that for the recent problem with hangs in the case of some data CRC
    errors, a udelay(10) inserted at the start of send_stop_abort() greatly
    helped in triggering the error, but I've not included this as part of
    the fault injection support since it seemed too specific.
    
    Signed-off-by: Vincent Whitchurch <vincent.whitchurch@axis.com>
    Reviewed-by: Jaehoon Chung <jh80.chung@samsung.com>
    Link: https://lore.kernel.org/r/20210701080534.23138-1-vincent.whitchurch@axis.com
    Signed-off-by: Ulf Hansson <ulf.hansson@linaro.org>

diff --git a/drivers/mmc/host/dw_mmc.h b/drivers/mmc/host/dw_mmc.h
index da5923a92e60..ce05d81477d9 100644
--- a/drivers/mmc/host/dw_mmc.h
+++ b/drivers/mmc/host/dw_mmc.h
@@ -14,6 +14,8 @@
 #include <linux/mmc/core.h>
 #include <linux/dmaengine.h>
 #include <linux/reset.h>
+#include <linux/fault-inject.h>
+#include <linux/hrtimer.h>
 #include <linux/interrupt.h>
 
 enum dw_mci_state {
@@ -230,6 +232,11 @@ struct dw_mci {
 	struct timer_list       cmd11_timer;
 	struct timer_list       cto_timer;
 	struct timer_list       dto_timer;
+
+#ifdef CONFIG_FAULT_INJECTION
+	struct fault_attr	fail_data_crc;
+	struct hrtimer		fault_timer;
+#endif
 };
 
 /* DMA ops for Internal/External DMAC interface */